### PR TITLE
SQC-652 Implement Hyperdrive TLS proxy in Miniflare

### PR DESCRIPTION
**Pull request by Ltadrian**
Fixes SQC-652.

This fixes the problem of when a user runs 
`npx wrangler dev` 
with a Hyperdrive binding and `localConnectionString` pointing to a database that requires TLS. In such cases, the command will result in the following TLS error:
```
{ "error": "connection is insecure (try using `sslmode=require`)" }
```

This change adds a proxy layer to Miniflare and introduces a proxy service for Hyperdrive bindings. The proxy layer attempts to establish a TLS connection with the appropriate database based on the specified sslmode parameter.

This allows users to simply set `sslmode=require` in their `localConnectionString`, enabling their Hyperdrive binding to automatically upgrade to TLS instead of requiring them to manually configure TLS settings in their database driver. (Note: this still will not work without additional changes to the binding service.)

---

- Tests
  - [X] Tests included
  - [ ] Tests not necessary because:
- Public documentation
  - [X] Cloudflare docs PR(s): https://github.com/cloudflare/cloudflare-docs/pull/26545
  - [ ] Documentation not necessary because:
- Wrangler V3 Backport
  - [ ] Wrangler PR: 
  - [X] Not necessary because: New feature behavior

### CODE REVIEW

File: packages/miniflare/src/plugins/hyperdrive/index.ts [hunk 1]
```diff
@@ -115,15 +115,56 @@ export const HYPERDRIVE_PLUGIN: Plugin<typeof HyperdriveInputOptionsSchema> = {
...
+				url.hostname,
+				getPort(url),
+				scheme,
+				parseSslMode(url, scheme)
+			);
+			services.push({
 				name: `${HYPERDRIVE_PLUGIN_NAME}:${name}`,
 				external: {
-					address: `${url.hostname}:${getPort(url)}`,
+					address: `localhost:${proxyPort}`,
```
Comment (petebacondarwin) [inline packages/miniflare/src/plugins/hyperdrive/index.ts:133]: would it be safer to run the proxy only on 127.0.0.1 and use that here rather than localhost?

File: packages/miniflare/src/plugins/hyperdrive/index.ts [hunk 2]
```diff
@@ -115,15 +115,56 @@ export const HYPERDRIVE_PLUGIN: Plugin<typeof HyperdriveInputOptionsSchema> = {
...
+			sslMode === "DISABLED" ||
+			sslMode === "false" ||
+			sslMode === "0"
+		) {
+			return "disable";
+		} else if (sslMode === "PREFERRED") {
+			return "prefer";
+		}
+
+		return "prefer"; // Default
```
Comment (petebacondarwin) [inline packages/miniflare/src/plugins/hyperdrive/index.ts]: This line is not really needed given that we return `prefer` directly below.

File: packages/miniflare/src/plugins/shared/index.ts [hunk 3]
```diff
@@ -22,6 +22,7 @@ import {
 	QueueProducerSchema,
 	sanitisePath,
 } from "../../workers";
+import { HyperdriveProxyController } from "../hyperdrive/hyperdrive-proxy";
```
Comment (petebacondarwin) [inline packages/miniflare/src/plugins/shared/index.ts]: Should this be a type import?


```suggestion
import type { HyperdriveProxyController } from "../hyperdrive/hyperdrive-proxy";
```

File: packages/wrangler/e2e/get-platform-proxy.test.ts [hunk 4]
```diff
@@ -365,70 +367,210 @@ describe("getPlatformProxy()", () => {
...
+			});
+		});
 
-			await seed(root, {
-				"wrangler.toml": dedent`
-						name = "hyperdrive-app"
-						compatibility_date = "2024-08-20"
-						compatibility_flags = ["nodejs_compat"]
+		it.skipIf(
+			// in CI this test fails for windows because of ECONNRESET issues
```
Comment (petebacondarwin) [inline packages/wrangler/e2e/get-platform-proxy.test.ts:407]: Do we know why this is?
Is it just a problem with the tests or is it a problem with Miniflare?

File: packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts [hunk 5]
```diff
@@ -1069,14 +1081,81 @@ describe("LocalRuntimeController", () => {
...
-				await writer.write(new TextEncoder().encode("👋"));
-				await writer.close();
-				return new Response(socket.readable);
-			}
-		}`);
+				async fetch(request, env, ctx) {
+					const socket = env.DB.connect();
+					const writer = socket.writable.getWriter();
+					await writer.write(new TextEncoder().encode("👋"));
+					writer.releaseLock();
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts]: How come this no longer closes the writer?

File: packages/miniflare/src/plugins/hyperdrive/index.ts (hunk 1 above)
Comment (Ltadrian) [reply packages/miniflare/src/plugins/hyperdrive/index.ts:133]: I've been testing this locally and so far for me using `localhost` works whereas setting `127.0.0.1` seems to fail to connect to the integration test server. Not sure exactly why but maybe there is some IPv4/IPv6 conflict?

Comment (petebacondarwin) [reply packages/miniflare/src/plugins/hyperdrive/index.ts:133]: Cool. No probs. Thanks for confirming

**Comment by Ltadrian** [comment]
> A few questions and suggestions. But it looks like this is a legitimate failure in CI? https://github.com/cloudflare/workers-sdk/actions/runs/19313849941/job/55240551979?pr=11219#step:6:3822

Yes this actually was a failure, I had missed this test but should be fixed now

File: packages/wrangler/e2e/get-platform-proxy.test.ts (hunk 4 above)
Comment (Ltadrian) [reply packages/wrangler/e2e/get-platform-proxy.test.ts:407]: I don't exactly know either, this Windows check was here originally and I just copied it over to all the Hyperdrive tests. I'll test this out on my Windows machine later today to see if it still fails

File: packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts (hunk 5 above)
Comment (Ltadrian) [reply packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts]: Closing the writer stream early was causing the proxy to not write back the expected 👋  response. I have updated this to remove the `writer.releaseLock()` and move `writer.close()` after the read and that seems to be working ok

File: packages/wrangler/e2e/get-platform-proxy.test.ts (hunk 4 above)
Comment (Ltadrian) [reply packages/wrangler/e2e/get-platform-proxy.test.ts:407]: Testing this on my windows machine and I am able to reproduce the ECONNRESET errors, 
```
Error: read ECONNRESET

at TCP.onStreamRead (node:internal/stream_base_commons:216:20)
Emitted 'error' event on Socket instance at:
  at emitErrorNT (node:internal/streams/destroy:170:8)
  at emitErrorCloseNT (node:internal/streams/destroy:129:3)
  at process.processTicksAndRejections (node:internal/process/task_queues:89:21) {
    errno: -4077,
    code: 'ECONNRESET',
    syscall: 'read'
}
```

**Comment by Ltadrian** [comment]
@petebacondarwin think I could get a re-review on this? I fixed some more tests and cleaned up some issues when testing more edge cases. I think currently the tests failing are flaky tests?

File: fixtures/vitest-pool-workers-examples/hyperdrive/src/index.ts [hunk 6]
```diff
@@ -4,7 +4,14 @@ export default <ExportedHandler<Env>>{
...
 
 		const socket = env.ECHO_SERVER_HYPERDRIVE.connect();
-		await request.body.pipeTo(socket.writable);
-		return new Response(socket.readable);
+		const writer = socket.writable.getWriter();
+
+		// parse body
+		let value = await request.text();
+		await writer.write(new TextEncoder().encode(value));
+		let result = await socket.readable.getReader().read();
```
Comment (vicb) [inline fixtures/vitest-pool-workers-examples/hyperdrive/src/index.ts]:
```suggestion
		const value = await request.text();
		await writer.write(new TextEncoder().encode(value));
		const result = await socket.readable.getReader().read();
```

File: fixtures/vitest-pool-workers-examples/hyperdrive/global-setup.ts [hunk 7]
```diff
@@ -2,12 +2,23 @@ import assert from "node:assert";
...
 import util from "node:util";
-import type { GlobalSetupContext } from "vitest/node";
+import type { TestProject } from "vitest/node";
 
 // Global setup runs inside Node.js, not `workerd`
-export default async function ({ provide }: GlobalSetupContext) {
+export default async function ({ provide }: TestProject) {
 	// Start echo server on random port
-	const server = net.createServer((socket) => socket.pipe(socket));
+	const sslRequestPacket = Buffer.from([
```
Comment (vicb) [inline fixtures/vitest-pool-workers-examples/hyperdrive/global-setup.ts]: Could you please add a comment on what this is?

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 8]
```diff
@@ -0,0 +1,349 @@
+import net from "node:net";
+import tls from "node:tls";
+
+export class HyperdriveProxyController {
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:34]: Could you please add JSDoc to class and methods.
(Could be a one liner)

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+	// Map hyperdrive binding name to proxy server
+	#servers = new Map<string, net.Server>();
+
+	async createProxyServer(
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: Maybe use an object to name parameters?

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+	return new Promise((resolve) => {
+		socket.once("data", (data) => resolve(data));
+	});
+}
+
+function tlsConnect(options: tls.ConnectionOptions): Promise<tls.TLSSocket> {
+	return new Promise((resolve, reject) => {
+		const socket = tls.connect(options);
+
+		socket.once("secureConnect", () => {
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: should you cancel the success listener on error and vice versa?

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+	// Skip null terminator
+	offset++;
+
+	// Ensure ther are enough bytes left for fixed fields
+	if (offset + 4 + 8 + 1 + 2 > payloadLength) return false;
+
+	// Skip connection_id, auth_plugin_data_part_1, filler
+	offset += 4;
+	offset += 8;
+	offset += 1;
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: Use constants with comments instead of magic numbers

**Comment by Ltadrian** [comment]
@ReppCodes @knickish @prydt could I get team approvals please 😄

File: fixtures/vitest-pool-workers-examples/hyperdrive/global-setup.ts [hunk 12]
```diff
@@ -2,12 +2,25 @@ import assert from "node:assert";
...
+		0x00, 0x00, 0x00, 0x08, 0x04, 0xd2, 0x16, 0x2f,
+	]);
 	// Start echo server on random port
-	const server = net.createServer((socket) => socket.pipe(socket));
+	const server = net.createServer((socket) => {
+		socket.on("data", (chunk) => {
+			// on postgres ssl request packet respond with 'N' to indicate no SSL support
+			if (sslRequestPacket.equals(chunk)) {
+				socket.write("N");
```
Comment (ReppCodes) [inline fixtures/vitest-pool-workers-examples/hyperdrive/global-setup.ts:18]: Is advertising no SSL support going to cause issues with sslmode=require?

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+		scheme: string,
+		sslmode: string
+	) {
+		// Connect to real database
+		const dbSocket = net.connect({ host: targetHost, port: targetPort });
+		const sslmodeRequire = sslmode === "require";
+		const sslmodePrefer = sslmode === "prefer";
+		if (sslmodePrefer || sslmodeRequire) {
+			try {
+				if (scheme === "postgres" || scheme === "postgresql") {
```
Comment (ReppCodes) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: Might be nice to have the supported schemes as a const array of magic up top. I doubt we'll add more, but for the sake of maintainability might be good to do.

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+/** Handles Postgres TLS connection */
+async function handlePostgresTlsConnection(
+	dbSocket: net.Socket,
+	clientSocket: net.Socket,
+	targetHost: string,
+	targetPort: number,
+	sslmodeRequire: boolean
+) {
+	// Send Postgres sslrequest bytes
+	const sslRequestPacket = Buffer.from([
```
Comment (ReppCodes) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: Ditto the "magic const" comment for this. This is tricky enough that having it duplicated around is not ideal.

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 15]
```diff
@@ -0,0 +1,419 @@
...
+		targetPort,
+		clientSocket
+	);
+	// Pipe plain TCP sockets
+	clientSocket.pipe(newDbSocket);
+	newDbSocket.pipe(clientSocket);
+}
+
+/** Handles MySQL TLS connection */
+async function handleMsqlTlsConnection(
```
Comment (ReppCodes) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: nit: `mysql`

File: packages/miniflare/src/plugins/hyperdrive/index.ts [hunk 16]
```diff
@@ -115,15 +115,63 @@ export const HYPERDRIVE_PLUGIN: Plugin<typeof HyperdriveInputOptionsSchema> = {
...
 				},
-			})
-		);
+			});
+		}
+		return services;
 	},
 };
+
+function parseSslMode(url: URL, scheme: string): string {
```
Comment (ReppCodes) [inline packages/miniflare/src/plugins/hyperdrive/index.ts:144]: Docs links for pg/mysql would be good here.

File: packages/wrangler/e2e/dev.test.ts [hunk 17]
```diff
@@ -588,9 +588,23 @@ describe.each([{ cmd: "wrangler dev" }])(
 
 describe("hyperdrive dev tests", () => {
 	let server: nodeNet.Server;
+	const sslRequestPacket = Buffer.from([
```
Comment (ReppCodes) [inline packages/wrangler/e2e/dev.test.ts]: Ditto magic constant comment here.

**Comment by ReppCodes** [comment]
Few minor comments, but LGTM!

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts (hunk 14 above)
Comment (Ltadrian) [reply packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: Tried creating a shared constant but it's a bit hard to do that in this repo with disjoint modules. I've created the shared const in a few places, miniflare/wrangle e2e/vitest-pool-workers-example modules

File: fixtures/vitest-pool-workers-examples/hyperdrive/global-setup.ts (hunk 12 above)
Comment (Ltadrian) [reply fixtures/vitest-pool-workers-examples/hyperdrive/global-setup.ts:18]: For integration tests/examples it should be fine to return no SSL support. In this case it should cause a failure for `sslmode=require`

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 18]
```diff
@@ -0,0 +1,427 @@
...
+				targetHost,
+				Number.parseInt(targetPort),
+				scheme,
+				sslmode
+			);
+		});
+		const port = await new Promise<number>((resolve) => {
+			server.listen(0, "localhost", () => {
+				const addr = server.address() as net.AddressInfo;
+				resolve(addr.port);
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: nit/suggestion
```suggestion
				const { port } = server.address() as net.AddressInfo;
				resolve(port);
```

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 19]
```diff
@@ -0,0 +1,427 @@
...
+			}
+		}
+		// Pipe plain tcp sockets
+		clientSocket.pipe(dbSocket);
+		dbSocket.pipe(clientSocket);
+	}
+
+	/** Disposes of the proxy servers when shutting down the worker.*/
+	async dispose(): Promise<void> {
+		await Promise.all(
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: `all` will reject as soon as one of the Promise rejects.
`allSettled` will wait for all Promises to settle (either successfully or not).

```suggestion
		await Promise.allSettled(
```

Would there be any value in adding a debug log on error?

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 20]
```diff
@@ -0,0 +1,427 @@
...
+		clientSocket
+	);
+
+	// Pipe plain TCP sockets
+	clientSocket.pipe(newDbSocket);
+	newDbSocket.pipe(clientSocket);
+	return;
+}
+
+/** Helper function to create and wait for plain TCP connection */
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]:
```suggestion
/** Create and wait for plain TCP connection */
```

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 21]
```diff
@@ -0,0 +1,427 @@
...
+
+	// Set up error handler
+	dbSocket.on("error", () => {
+		clientSocket.destroy();
+	});
+
+	return dbSocket;
+}
+
+/** Helper function to set up TLS connection with pipes and error handlers */
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]:
```suggestion
/** Set up TLS connection with pipes and error handlers */
```

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 22]
```diff
@@ -0,0 +1,427 @@
...
+		clientSocket.destroy();
+		tlsSocket.destroy();
+	});
+
+	// Pipe sockets
+	clientSocket.pipe(tlsSocket);
+	tlsSocket.pipe(clientSocket);
+}
+
+/** Helper functions for net library */
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: Nit: Could the comment described what the function is actually doing?

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 23]
```diff
@@ -0,0 +1,427 @@
...
+function writeAsync(
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:346]: suggestion, it looks like this function is often called with `Buffer.from(array)`, would it makes sense that bytes could also be an array and turned into a Buffer here if needed?

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 24]
```diff
@@ -0,0 +1,427 @@
...
+			MYSQL_FILLER_LENGTH +
+			MYSQL_CAPABILITY_FLAGS_LOWER_LENGTH >
+		payloadLength
+	)
+		return false;
+
+	// Skip connection_id, auth_plugin_data_part_1, filler
+	offset += MYSQL_CONNECTION_ID_LENGTH;
+	offset += MYSQL_AUTH_PLUGIN_DATA_PART_1_LENGTH;
+	offset += MYSQL_FILLER_LENGTH;
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:415]: Would it make sense to move those lines up to line 405 to remove the duplication in the if condition ?

File: packages/miniflare/src/plugins/hyperdrive/index.ts [hunk 25]
```diff
@@ -115,15 +115,66 @@ export const HYPERDRIVE_PLUGIN: Plugin<typeof HyperdriveInputOptionsSchema> = {
...
-		);
+			});
+		}
+		return services;
 	},
 };
+
+// Postgres sslmode docs: https://www.postgresql.org/docs/current/libpq-ssl.html
+// MySQL ssl-mode docs: https://dev.mysql.com/doc/refman/8.4/en/using-encrypted-connections.html
+function parseSslMode(url: URL, scheme: string): string {
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/index.ts:144]: Question: should search params always be normalized (to either casing)?

It's only done for some code paths.

File: packages/miniflare/src/plugins/hyperdrive/index.ts [hunk 26]
```diff
@@ -115,15 +115,66 @@ export const HYPERDRIVE_PLUGIN: Plugin<typeof HyperdriveInputOptionsSchema> = {
...
+	if (scheme === "postgres" || scheme === "postgresql") {
+		return url.searchParams.get("sslmode") || "disable"; // disable is default
+	} else if (scheme === "mysql") {
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/index.ts:155]: nit: I find `else`s after a returning branch harder to read but I understand it can be personal preferences

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts [hunk 27]
```diff
@@ -0,0 +1,427 @@
...
+				clientSocket,
+				targetHost,
+				Number.parseInt(targetPort),
+				scheme,
+				sslmode
+			);
+		});
+		const port = await new Promise<number>((resolve) => {
+			server.listen(0, "localhost", () => {
+				const addr = server.address() as net.AddressInfo;
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts]: Should the type be tested here?
It is done [in this PR](https://github.com/cloudflare/workers-sdk/pull/11219/files#diff-613be7b66345d5bbec95c7691cdeea693e1dbaa7418e7043faefb41d90cea20cR395)

File: packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts [hunk 28]
```diff
@@ -1047,9 +1047,21 @@ describe("LocalRuntimeController", () => {
...
+			]);
+			const server = net.createServer((socket) => {
+				socket.on("data", (chunk) => {
+					if (POSTGRES_SSL_REQUEST_PACKET.equals(chunk)) {
+						socket.write("N");
+					} else {
+						// echo back result
+						socket.write(chunk);
+					}
+				});
```
Comment (vicb) [inline packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts]: I feel like I have been seeing this code many times at this point in the review.

Is there an opportunity to factor some code ?

**Review by vicb — APPROVED** [review]
LGTM.

I have added mostly minor comments / suggestions.

If one thing, I feel like there is quite some code duplication across the PR.

File: packages/miniflare/src/plugins/hyperdrive/index.ts (hunk 26 above)
Comment (Ltadrian) [reply packages/miniflare/src/plugins/hyperdrive/index.ts:155]: agree it was a bit messy, updated to use ` switch ` statement

File: packages/miniflare/src/plugins/hyperdrive/index.ts (hunk 25 above)
Comment (Ltadrian) [reply packages/miniflare/src/plugins/hyperdrive/index.ts:144]: I agree, I have normalized param key/values to all lowercase

File: packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts (hunk 23 above)
Comment (Ltadrian) [reply packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:346]: I've updated this bytes param to take union type of `Buffer<ArrayBuffer> | Uint8Array | number[]`. I really enjoy this feature of typescript 😄

File: packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts (hunk 28 above)
Comment (Ltadrian) [reply packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts]: I have refactored these simple echo functions to a shared handler function `createPostgresEchoHandler`  https://github.com/cloudflare/workers-sdk/pull/11219/files#diff-618bb77d274502bf29afc0e51260735e6d860ac9a89e5155e501f2c9fb294734R11

File: packages/miniflare/src/plugins/hyperdrive/index.ts [hunk 29]
```diff
@@ -115,15 +115,68 @@ export const HYPERDRIVE_PLUGIN: Plugin<typeof HyperdriveInputOptionsSchema> = {
...
 };
+
+// Postgres sslmode docs: https://www.postgresql.org/docs/current/libpq-ssl.html
+// MySQL ssl-mode docs: https://dev.mysql.com/doc/refman/8.4/en/using-encrypted-connections.html
+function parseSslMode(url: URL, scheme: string): string {
+	// Normalize keys/values to lowercase
+	const params = Object.fromEntries(
+		Array.from(url.searchParams.entries()).map(([k, v]) => [
+			k.toLowerCase(),
+			v.toLocaleLowerCase(),
```
Comment (vicb) [inline packages/miniflare/src/plugins/hyperdrive/index.ts]: Do you really want the Locale version here?

Comment (Ltadrian) [reply packages/miniflare/src/plugins/hyperdrive/index.ts]: Thanks for catching typo 🫠

File: .changeset/cute-plums-relate.md [hunk 30]
```diff
@@ -0,0 +1,7 @@
+---
+"miniflare": minor
+"wrangler": minor
+---
+
+Implement Hyperdrive binding TLS miniflare proxy. This will allow for wrangler dev hyperdrive bindings to connect to external
+databases that require TLS.
```
Comment (dario-piotrowicz) [inline .changeset/cute-plums-relate.md:7]: nit: 
```suggestion
Implement Hyperdrive binding TLS miniflare proxy. This will allow for wrangler dev hyperdrive bindings to connect to external databases that require TLS.
```
